tests/test_object_store_service.py

```diff
@@ -193,6 +195,34 @@ def test_generate_url(self):
                         " access generated url")
                 self.assertEqual(requests.get(url).content, content)
 
+    @helpers.skipIfNoService(['storage.buckets'])
+    def test_generate_url_write_permissions(self):
+        name = "cbtestbucketobjs-{0}".format(helpers.get_uuid())
+        test_bucket = self.provider.storage.buckets.create(name)
+
+        with cb_helpers.cleanup_action(lambda: test_bucket.delete()):
+            obj_name = "hello_upload_download.txt"
+            obj = test_bucket.objects.create(obj_name)
+
+            with cb_helpers.cleanup_action(lambda: obj.delete()):
+                content = b"Hello World. Generate a url."
+
+                url = obj.generate_url(100, writable=True)
+                if isinstance(self.provider, TestMockHelperMixin):
+                    raise self.skipTest(
+                        "Skipping rest of test - mock providers can't"
+                        " access generated url")
+                elif isinstance(self.provider, GCPCloudProvider):
+                    requests.put(url, data=content)
+                elif isinstance(self.provider, AWSCloudProvider):
+                    requests.put(url['url'], data=url['fields'], files={"file": (obj_name, content)})
+                else:
+                    requests.post(url, data=content)
```

Comment:
```suggestion
                else:
                    requests.put(url, data=content)
```

Reply:
@FabioRosado I was just trying to follow why the AWS provider's request.put has more parameters? The docs suggest that `requests.put(url, data=content)` should also work?

Reply:
That's because AWS doesn't return a single URL like GCP. It returns a dictionary containing: `url - endpoint` and `fields - all the needed data to get things uploaded`.

This is the [docs that I was following](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-presigned-urls.html?highlight=generate_presigned_url#generating-a-presigned-url-to-upload-a-file) were you looking at the same ones?

Reply:
@FabioRosado I was actually looking at this: https://docs.aws.amazon.com/code-samples/latest/catalog/python-s3-s3_basics-presigned_url.py.html
I guess the core problem here is that we'll need to have a consistent return value across all providers, so that cloud agnosticism is maintained. If the above code sample doesn't work, maybe the solution is to always return a dict instead? On everything other than AWS, `fields`, and `file` would be empty?

Reply:
I agree with you that its not good that we can't have a consistent return value across providers. The annoying thing with AWS is that they want these fields in the body of the request while the others give you a single url that you can use.

We could change it as you suggest, but wouldn't be confusing that a get method returns a string and a post one returns a dict?

Reply:
@FabioRosado It looks like using `generate_presigned_url` would make this consistent?
https://stackoverflow.com/questions/65198959/aws-s3-generate-presigned-url-vs-generate-presigned-post-for-uploading-files

Reply:
These changes make the tests pass as expected: https://github.com/CloudVE/cloudbridge/pull/295